scripts/cbmc-viewer/block.py

```diff
@@ -0,0 +1,275 @@
+"""Basic block coverage reporting for CBMC."""
+
+from builtins import object
+import re
+import os
+from pprint import pprint
+
+import linestatus
+import srcloc as SrcLoc
+
+class Block(object):
+    """Basic block coverage for CBMC.
+
+    The Block class parses the basic block coverage information in the
+    xml file produced by 'cbmc --cover locations --xml-ui' and
+    provides three classes of coverage information:
+
+        block.line[file][line] is the line status HIT, PARTIAL, MISS,
+        or None giving whether this line in this file was hit or not.
+
+        block.file[file] gives the line coverage for a file in the
+        form of a tuple (coverage, hit, reachable) giving the number
+        of reachable lines, the number of hit lines, and their ratio
+        coverage.
+
+        block.function[file][func] gives the line coverage for a
+        function in a file in the form of a tuple (coverage, hit,
+        reachable) giving the number of reachable lines, the number of
+        hit lines, and their ratio coverage.
+    """
+
+    def __init__(self, xml="", srcdir="", blddir="",
+                 reachable=None, srcloc=None):
+        """Initialize block coverage statistics.
+
+        Args:
+
+           xml (str): the name of the xml file containing the block
+           coverage data produced by cbmc.
+
+           srcdir (str): the root of the source directory
+
+           blddir (str): the root of the source directory used to
+           build the goto binary (which may be different from srcdir
+           if the source directory was preprocessed before building)
+
+           reachable (obj): an instance of the Reachable class giving
+           statistics on the statically reachable functions
+        """
+
+        # pylint: disable=too-many-arguments
+
+        self.line = {}
+        self.function = {}
+        self.file = {}
+
+        # Record the list of reachable functions for later use
+        self.reachable = reachable.functions if reachable else []
+        # Record the mapping of filenames computed below for later use
+        # This should be using srcloc.clean_path
+        self.cleanup_filename = Block.cleanup_filename_function(blddir or
+                                                                srcdir)
+
+        if not xml:
+            return
```

Comment:
Also, my understanding is that quite a few functions here have failure modes that are not being handled properly, and they just fail silently. Is this what should happen?

Reply:
We have added a generic exception handler to cbmc-viewer.  We have added improved exception handling to the TODO.md list of things to fix in the future.